x/bank/keeper/keeper.go

```diff
@@ -412,6 +417,58 @@ func (k BaseKeeper) BurnCoins(ctx context.Context, moduleName string, amounts sd
 	return nil
 }
 
+// SubtractCoins subtracts coins from the balance of an account
+// NOTE(Heimdall-v2): required in heimdall for fee withdrawal in topup module
+func (k BaseKeeper) SubtractCoins(ctx context.Context, moduleName string, addr sdk.AccAddress, amounts sdk.Coins) error {
+	if !amounts.IsValid() {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidCoins, amounts.String())
+	}
+
+	err := k.subUnlockedCoins(ctx, addr, amounts)
+	if err != nil {
+		return err
+	}
+
+	for _, amount := range amounts {
+		supply := k.GetSupply(ctx, amount.GetDenom())
+		supply = supply.Sub(amount)
+		k.setSupply(ctx, supply)
+	}
+
+	k.logger.Debug("subtracted tokens from account", "amount", amounts.String(), "from", moduleName)
+
+	// emit burn event
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	sdkCtx.EventManager().EmitEvent(
+		types.NewCoinBurnEvent(addr, amounts),
+	)
+
+	return nil
+}
+
+// SetCoins sets the balance of an account to a given amount
+// NOTE(Heimdall-v2): could be needed
+func (k BaseKeeper) SetCoins(ctx context.Context, addr sdk.AccAddress, amounts sdk.Coins) error {
```

Comment:
If this isn't used anywhere why implement?

Reply:
It is being used in heimdall in some places like [this](https://github.com/maticnetwork/heimdall/blob/249aa798c2f23c533d2421f2101127c11684c76e/app/app.go#L164)so decided to add it here. We can remove it later if we find an alternate way to achieve this OOTB.

Reply:
I suspect this might be an antipattern but we discussed this already wrt switching to expected keepers

Reply:
Even if we switch to expected keepers interface, the implementation of this method has to still reside in bank right ?

Reply:
Not sure if done properly, but in `gov` PR we have something similar:
- here we [have](https://github.com/0xPolygon/cosmos-sdk/pull/7/files#diff-f4b8a72fd7caee61e80e763102ad4578109875ddace08521cda0efe4eadae1d3R33) an expected keeper method for the `StakingKeeper` interface 
- [here](https://github.com/0xPolygon/cosmos-sdk/pull/7/files#diff-bd9428f4e6e0c1385c6122f03a090b2ed52aa6bae9ea4a896b61917e5a858796R186) I implemented a "fake" version of it, required to build the package
- the real implementation of such method, used by the `gov` module, will then reside in the staking module (@VAIBHAVJINDAL3012 is working on it, but I believe it's [this one](https://github.com/0xPolygon/heimdall-v2/blob/vaibhav/StakeModule/x/stake/keeper/validator.go#L446)

Reply:
Hmm, this also seems feasible. Although to me it seems more appropriate that the implementation is done in bank keeper. Thoughts @glnro ?

Reply:
Yeah not sure about this method specifically. In `gov` 's case, the expected implementation is in staking module, so I believe that's the right thing to do.

Reply:
Leaving it to Lauren for further clarifications.

Reply:
What is the use case for this?

Reply:
So I went through again where `SetCoins` is being used and apart from some tests I didn't see it being used much.
Maybe we can remove this after all and later if the need arises, add it back instead.

Reply:
Removed [1a3e51f](https://github.com/0xPolygon/cosmos-sdk/pull/5/commits/1a3e51f0a7a0f3a76d0b4b83b9bbd85101280916).